HP9845: added synchronous I/O for RS232 port and HP98046 module

fulivi opened the pull request:
Hi,

this PR reopens #7528 with one major enhancement: I updated the HP98046 module to use the new F4702 BRG device.
Note that I had to change the `f4702_device::im_w` function to ensure that external clocks are propagated even when the F4702 device is not executing. This change is needed for HP98046 because it has 2 cascaded BRGs.

Thanks.
--F.Ulivi

ajrhacker commented:
The F4702 change is wrong. The device executes continuously regardless of the S input state, so IM will propagate to Z upon the next master clock cycle (the Z output is always thus synchronized).

fulivi commented:
The way I see it (and I tested it on HP98046) is that if one BRG is 
outputting a clock that goes into the Im input of the other BRG, the 
clock is not propagating through the second one for the very simple 
reason that at any given time there's just one device executing. So if 
the first BRG, in one execution timeslice, outputs more than 1 period of 
clock the device that should receive it through the second BRG is seeing 
nothing. Am I missing something? Maybe there's a simpler way to do it.

ajrhacker commented:
I think I see what's going on. The cross-connection of one BRG's Z output to the other's IM input is highly unusual, and here occurs only in loopback mode.

You may need to add a synchronization callback instead of calling `machine().scheduler().synchronize()` with no parameters.

fulivi commented:
Hi,
can you explain a little bit what you mean when you suggest adding a synchronization callback and why would that be an improvement, please? I'm not a super-expert of the scheduling core of MAME, I just know that the current code works.
Thanks.

ajrhacker commented:
At least the im_w calls here needs to be moved to within callbacks so they will be scheduled properly.

cuavas commented on src/devices/bus/rs232/sync_io.cpp:
```diff
@@ -0,0 +1,213 @@
+	m_clk_timer->adjust(period , 0 , period);
+}
+
+DEFINE_DEVICE_TYPE(SYNC_IO, sync_io_device, "sync_io", "RS232 Synchronous I/O")
```
The class and device names are too generic.  Please use something more specific.

fulivi commented:
I changed the name of the synchronous port to `rs232_sync_io`.
I also moved the clock input from F4702s to use synchronized functions. Hopefully I got the suggestion from AJR right. 
Now, do you think there's anything else I need to change, please?
Thanks.

ajrhacker commented on src/devices/machine/f4702.cpp:
```diff
@@ -152,6 +152,12 @@ void f4702_device::reset_counters()
 WRITE_LINE_MEMBER(f4702_device::im_w)
 {
 	m_im = state;
+
+	if (m_s == 0 || m_s == 1)
+	{
+		// Ensure Im input is propagated even when device is not "executing"
+		m_z_callback(m_main_counter & 0x007, m_im);
+	}
```
This is the part that isn't wanted. Is it still needed? (It's technically incorrect since the real device propagates Im synchronously even after multiplexing.)

fulivi commented:
At last! This one works and doesn't require any change to F4702 driver.
